### PR TITLE
feat: add volume group API endpoint interaction

### LINKED ISSUE #86: Feature Request: Add LVM Interfaces
The client supports RAID and BCache sets but does not seem to handle LVM (or ZFS) layouts. There are currently requests up on the TF provider issue tracker to handle these resources/data types which would first mandate implementation at this layer.

### PR DESCRIPTION
**Pull request by summerwind**
This PR adds the support of following MAAS API endpoint.

- `DELETE /MAAS/api/2.0/nodes/{system_id}/volume-group/{id}/`: Delete volume group
- `GET /MAAS/api/2.0/nodes/{system_id}/volume-group/{id}/`: Read a volume group
- `PUT /MAAS/api/2.0/nodes/{system_id}/volume-group/{id}/`: Update a volume group
- `POST /MAAS/api/2.0/nodes/{system_id}/volume-group/{id}/op-create_logical_volume`: Create a logical volume
- `POST /MAAS/api/2.0/nodes/{system_id}/volume-group/{id}/op-delete_logical_volume`: Delete a logical volume
- `GET /MAAS/api/2.0/nodes/{system_id}/volume-groups/`: List all volume groups
- `POST /MAAS/api/2.0/nodes/{system_id}/volume-groups/`: Create a volume group

I think this change would resolve #86.

### CODE REVIEW

**Comment by summerwind** [comment]
I'm trying to fill out to the CLA form, but not sure what to enter the field '_Please add the Canonical Project Manager or contact_'.

**Comment by skatsaounis** [comment]
Hi @summerwind,

Thank you for the contribution. You can put `N/A` for that box or my name and proceed. Please let me know when this is done or if you face any other issue.

**Comment by summerwind** [comment]
@skatsaounis Thank you for following up!
I have signed the CLA form.

**Comment by skatsaounis** [comment]
Hi @summerwind. Thank you for signing the CLA. As you observed already, I re-triggered the actions and now everything looks green. Please give us some time to review your PR and if any change is needed we will come back to you. Thank you for your patience :slightly_smiling_face:

File: entity/volume_group.go [hunk 1]
```diff
@@ -2,17 +2,44 @@ package entity
...
+	HumanUsedSize      string          `json:"human_used_size,omitempty"`
+	Name               string          `json:"name,omitempty"`
+	LogicalVolumes     []LogicalVolume `json:"logical_volumes,omitempty"`
+	Size               int64           `json:"size,omitempty"`
+	UsedSize           int64           `json:"used_size,omitempty"`
+	AvailableSize      int64           `json:"available_size,omitempty"`
+	ID                 int             `json:"id,omitempty"`
+}
+
+// VolumeGroupParams enumerates the parameters for the volume group operation
```
Comment (skatsaounis) [inline entity/volume_group.go:20]: You should use a different Params struct for update. Copied from the source code [link-1](https://github.com/canonical/maas/blob/master/src/maasserver/forms/__init__.py#L4213) [link-2](https://github.com/canonical/maas/blob/master/src/maasserver/api/volume_groups.py#L209):

```py
name = forms.CharField(required=False)
uuid = UUID4Field(required=False)
add_block_devices = forms.MultipleChoiceField(required=False)
remove_block_devices = forms.MultipleChoiceField(required=False)
add_partitions = forms.MultipleChoiceField(required=False)
remove_partitions = forms.MultipleChoiceField(required=False)
```
So we need something like:

```go
type VolumeGroupUpdateParams struct {
	Name               string   `url:"name,omitempty"`
	UUID               string   `url:"uuid,omitempty"`
	AddBlockDevices    []string `url:"add_block_devices,omitempty"`
        RemoveBlockDevices []string `url:"remove_block_devices,omitempty"`
	AddPartitions      []string `url:"add_partitions,omitempty"`
        RemovePartitions   []string `url:"remove_partitions,omitempty"`
}
```

File: client/volume_group.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	volumeGroup := new(entity.VolumeGroup)
+	err := v.client(systemID, id).Get("", url.Values{}, func(data []byte) error {
+		return json.Unmarshal(data, volumeGroup)
+	})
+
+	return volumeGroup, err
+}
+
+// Update updates a given VolumeGroup
+func (v *VolumeGroup) Update(systemID string, id int, params *entity.VolumeGroupParams) (*entity.VolumeGroup, error) {
```
Comment (skatsaounis) [inline client/volume_group.go:33]:
```suggestion
func (v *VolumeGroup) Update(systemID string, id int, params *entity.VolumeGroupUpdateParams) (*entity.VolumeGroup, error) {
```

File: api/volume_group.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"github.com/canonical/gomaasclient/entity"
+)
+
+// VolumeGroup is an interface providing API behaviour for
+// volume groups
+type VolumeGroup interface {
+	Get(systemID string, id int) (*entity.VolumeGroup, error)
+	Update(systemID string, id int, params *entity.VolumeGroupParams) (*entity.VolumeGroup, error)
```
Comment (skatsaounis) [inline api/volume_group.go:11]:
```suggestion
	Update(systemID string, id int, params *entity.VolumeGroupUpdateParams) (*entity.VolumeGroup, error)
```

File: entity/volume_group.go [hunk 4]
```diff
@@ -2,17 +2,44 @@ package entity
...
-	ID                 int         `json:"id,omitempty"`
+	Devices            interface{}     `json:"devices,omitempty"`
+	ResourceURI        string          `json:"resource_uri,omitempty"`
+	HumanSize          string          `json:"human_size,omitempty"`
+	UUID               string          `json:"uuid,omitempty"`
+	HumanAvailableSize string          `json:"human_available_size,omitempty"`
+	SystemID           string          `json:"system_id,omitempty"`
+	HumanUsedSize      string          `json:"human_used_size,omitempty"`
+	Name               string          `json:"name,omitempty"`
+	LogicalVolumes     []LogicalVolume `json:"logical_volumes,omitempty"`
```
Comment (skatsaounis) [inline entity/volume_group.go:13]: Why not using `BlockDevice`? or even better create a `VirtualBlockDevice` which is missing from our entities and is an extension of a `BlockDevice`? [link-1](https://github.com/canonical/maas/blob/master/src/maasserver/api/volume_groups.py#L175) [link-2](https://github.com/canonical/maas/blob/master/src/maasserver/models/filesystemgroup.py#L871) [link-3](https://github.com/canonical/maas/blob/master/src/maasserver/models/virtualblockdevice.py#L65)

File: entity/volume_group.go [hunk 5]
```diff
@@ -2,17 +2,44 @@ package entity
...
+// VolumeGroupParams enumerates the parameters for the volume group operation
+type VolumeGroupParams struct {
+	Name         string   `url:"name,omitempty"`
+	UUID         string   `url:"uuid,omitempty"`
+	BlockDevices []string `url:"block_devices,omitempty"`
+	Partitions   []string `url:"partitions,omitempty"`
+}
+
+// LogicalVolume represents the LogicalVolume object.
+type LogicalVolume struct {
```
Comment (skatsaounis) [inline entity/volume_group.go:29]: Based on the above comment this struct can be deleted

**Review by skatsaounis — CHANGES_REQUESTED** [review]
Hi @summerwind . I would like to apologize for the big delay in the review and also to thank you once again for your contribution. Overall the PR looks good but I was able to spot 1-2 things that need to change. I am happy to merge when they are addressed :slightly_smiling_face:

File: entity/volume_group.go (hunk 1 above)
Comment (summerwind) [reply entity/volume_group.go]: Fixed in [1440eb0](https://github.com/canonical/gomaasclient/pull/94/commits/1440eb029a45da1bf36b1464eddaa2916e68c2a3).

File: entity/volume_group.go (hunk 4 above)
Comment (summerwind) [reply entity/volume_group.go]: I'm not sure if it is correct to use the `BlockDevice` type as is or to add `VirtualBlockDevice`. However using `BlockDevice` seems simpler and better at the moment, so I fixed it in [9464fad](https://github.com/canonical/gomaasclient/pull/94/commits/9464fadd3f0cbb0e9dc56e7606d52c180423e50e).

**Comment by summerwind** [comment]
@skatsaounis Thanks for the review!
I've fixed the code according to your feedback.

File: entity/volume_group.go (hunk 4 above)
Comment (summerwind) [reply entity/volume_group.go]: Would it be more appropriate to define it as `VirtualBlockDevice`? I will make corrections if necessary.

Comment (skatsaounis) [reply entity/volume_group.go]: That would be ideal, yes. The only reason I proposed the `BlockDevice` was to not overwhelm you. You could give it a try and I would propose any missing piece afterwards.

Comment (summerwind) [reply entity/volume_group.go]: I've added `VirtualBlockDevice` as an extension to `BlockDevice` in 374a4df. Could you please confirm that this is what you intended?

**Comment by summerwind** [comment]
CLA Check seems to have failed due to certificate expiration.
https://github.com/canonical/gomaasclient/actions/runs/11621308073/job/32364801159?pr=94

> Checking the following users on GitHub:
> Check in the signed list service
> Error occurred while checking user: certificate has expired

**Review by skatsaounis — APPROVED** [review]
LGTM +1 Thank you for your contribution and the patience @summerwind

**Comment by summerwind** [comment]
Thank you for the review!

**Pull request merged**